contract-derive/src/lib.rs

```diff
@@ -73,10 +73,38 @@ pub fn contract(_attr: TokenStream, item: TokenStream) -> TokenStream {
         }
     }).collect();
 
+    let emit_helper = quote! {
+        fn emit<T: alloy_sol_types::SolEvent>(event: T) {
+            use alloy_sol_types::private::B256;
+            
+            let log_data = event.encode_log_data();
+            let topics = event.encode_topics();
+            let num_topics = topics.len();
+            
+            let mut all_topics = [0u8; 96]; 
+            for (i, topic) in topics.iter().enumerate() {
+                if i >= 3 { break; } 
+                let start = i * 32;
+                let topic_bytes: B256 = (*topic).into();
+                all_topics[start..start + 32].copy_from_slice(topic_bytes.as_ref());
+            }
+    
+            log(
+                log_data.data.as_ptr() as u64,
+                log_data.data.len() as u64,
+                all_topics.as_ptr() as u64,
+                num_topics as u64  
+            );
+        }
+    };
```

Comment:
Nice! I think this can be broken down in 2 parts though:

the low level `log(...)` call could be extracted in its own library function that takes in the necessary generic arguments.

This rule here could be used to infer the event functions automatically for impl functions annotated with `#[event]` (eg),

Reply:
Yes!

Reply:
I just made the version you mentioned. If you’d like, I can upload everything with the complete code.

It works correctly

# contract-derive/lib.rs :

```rust
    let emit_helper = quote! {
        fn emit(event_name: &str, from: Address, to: Address, value: u64) {
            use alloy_sol_types::SolValue;
            use alloy_core::primitives::{keccak256, B256};
            
            let signature = [event_name.as_bytes(), b"(address,address,uint64)"].concat();
            let topic0 = B256::from(keccak256(&signature));
            
            let from_bytes = from.abi_encode();
            let to_bytes = to.abi_encode();
            let from_topic = B256::from_slice(&from_bytes);
            let to_topic = B256::from_slice(&to_bytes);
            
            let topics = [topic0, from_topic, to_topic];
            
            let value_data = value.abi_encode();
            
            eth_riscv_runtime::emit_log(&value_data, &topics);
        }
    };
```


# eth-riscv-runtime/log.rs

```rust
use alloy_core::primitives::B256;

pub fn emit_log(data: &[u8], topics: &[B256]) {
    let mut all_topics = [0u8; 96];
    for (i, topic) in topics.iter().enumerate() {
        if i >= 3 { break; }
        let start = i * 32;
        all_topics[start..start + 32].copy_from_slice(topic.as_ref());
    }

    crate::log(
        data.as_ptr() as u64,
        data.len() as u64,
        all_topics.as_ptr() as u64,
        topics.len() as u64
    );
}

```


# erc20

```rust
#[derive(Default)]
pub struct ERC20 {
    balance: Mapping<Address, u64>,
}

#[contract]
impl ERC20 {
    pub fn balance_of(&self, owner: Address) -> u64 {
        self.balance.read(owner)
    }

    pub fn transfer(&self, from: Address, to: Address, value: u64) -> bool {
        let from_balance = self.balance.read(from);
        let to_balance = self.balance.read(to);

        if from == to || from_balance < value {
            revert();
        }

        self.balance.write(from, from_balance - value);
        self.balance.write(to, to_balance + value);

        emit("Transfer", from, to, value);
        
        true
    }
```

Reply:
> I just made the version you mentioned. If you’d like, I can upload everything with the complete code.

@scab24 please just commit all your changes and we'll review it as part of the PR, it's easier than in comments.